### PR TITLE
Feat/arm tests

### PR DESCRIPTION
**Pull request by AbbeySieg**
added actual arm functionality

### CODE REVIEW

**Comment by AbbeySieg** [comment]
whoops those last three commits should be on a new branch for the claw, not this one

**Comment by AbbeySieg** [comment]
also last 5 commits are Daniel on Abbey's account as in right now

File: src/main/cpp/io/ArmIO.cpp [hunk 1]
```diff
@@ -32,23 +32,83 @@ bool ArmIO::ProcessIO() {
...
+            hw_interface_->arm_up_motor->Set(0);
+        } else {
+            // otherwise, it's cool
+            hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
+            hw_interface_->arm_up_motor->Set(-sw_interface_->arm_lift_power);
+        }
+    } else { // if neither limit is reached
+        hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
+        hw_interface_->arm_up_motor->Set(-sw_interface_->arm_lift_power);
+    }
```
Comment (jkleiber) [inline src/main/cpp/io/ArmIO.cpp:77]: On one hand this is a great use of the IO class because it guarantees hardware protection. On the other hand, the software subsystem is probably winding up the integrator of the PID if these limits are being hit. 

No change needed yet, but if control issues start appearing with a non-zero Ki, then we will need to send a boolean flag up to the software to tell it to stop adding to the integrator

File: src/main/cpp/io/ArmIO.cpp [hunk 2]
```diff
@@ -32,23 +32,83 @@ bool ArmIO::ProcessIO() {
...
+        if (sw_interface_->arm_lift_power > 0) {
+            // stop it
+            hw_interface_->arm_lift_motor->Set(0);
+            hw_interface_->arm_up_motor->Set(0);
+        } else {
+            // otherwise, it's cool
+            hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
+            hw_interface_->arm_up_motor->Set(-sw_interface_->arm_lift_power);
+        }
+    } else if (sw_interface_->arm_duty_cycle_encoder <= -110) {
```
Comment (jkleiber) [inline src/main/cpp/io/ArmIO.cpp:63]: Should we parameterize the +/- 110 in the TOML file?

File: src/main/cpp/io/ArmIO.cpp [hunk 3]
```diff
@@ -32,23 +32,83 @@ bool ArmIO::ProcessIO() {
...
+        if (sw_interface_->arm_extend_power > 0) {
+            // let it pass, so the arm can move outwards
+            hw_interface_->arm_extend_motor->Set(sw_interface_->arm_extend_power);
+        } else {
+            // otherwise, limit it to 0 and don't let it move
+            hw_interface_->arm_extend_motor->Set(0);
+        }
+    } else { // if the limit switch isn't pressed
+        // if we're instead at the farthest point
+        if (sw_interface_->arm_extend_encoder > 100) {
```
Comment (jkleiber) [inline src/main/cpp/io/ArmIO.cpp:96]: Should we parameterize this 100 in the TOML file? If not, then at the very least I think these magic numbers in this file need to be `#define`d in ArmIO.h so they have names associated with them

File: src/main/cpp/subsystems/Arm.cpp [hunk 4]
```diff
@@ -69,37 +102,22 @@ void Arm::Periodic() {
             VOKC_CALL(this->ManualControl());
             break;
         case Auto:
-            //this->interface_->arm_lift_power = this->arm_pid_->Caluculate(this->interface_->arm_encoder);
-            //this->interface_->arm_up_power = this->arm_pid_->Caluculate(this->interface_->arm_extend_encoder);
+            VOKC_CHECK(interface_ != nullptr);
+            VOKC_CHECK(this->arm_pid_ != nullptr);
+            this->interface_->arm_lift_power = this->arm_pid_->Calculate(this->interface_->arm_duty_cycle_encoder);
+            this->interface_->arm_extend_power = this->inches_pid_->Calculate(this->interface_->arm_extend_encoder);
```
Comment (jkleiber) [inline src/main/cpp/subsystems/Arm.cpp:108]: It would be best to make this block of code a separate function so the switch-case is clean

File: src/main/cpp/subsystems/Claw.cpp [hunk 5]
```diff
@@ -1,27 +1,81 @@
...
+bool Claw::Reset() {
+    this->ResetPositionEncoder();
+    this->ResetPositionPID();
+
+    return true;
 }
-void Claw::SimulationPeriodic(){
+
+void Claw::Periodic() {
+    interface_->claw_open_and_close_power = this->claw_pid_->Calculate(interface_->encoder_reading);
```
Comment (jkleiber) [inline src/main/cpp/subsystems/Claw.cpp:67]: maybe add a comment here

File: src/main/deploy/parameters.toml [hunk 6]
```diff
@@ -46,4 +46,24 @@ kD = 0.0
...
+
+[arm.lift_pid]
+kP = 0.1
+kI = 0
+kD = 0
+
+[arm.extend_pid]
+kP = 0.1
+kI = 0
+kD = 0
```
Comment (jkleiber) [inline src/main/deploy/parameters.toml:59]: I'm pretty surprised that only P control is being used here - does that work well?

File: src/main/include/RobotContainer.h [hunk 7]
```diff
@@ -121,6 +133,11 @@ class RobotContainer {
 
     //arm
     std::shared_ptr<ManualArmCommand> manual_arm_command_;
+
+    std::shared_ptr<IncrementArmExtendCommand> extendArmCommand;
+    std::shared_ptr<IncrementArmExtendCommand> retractArmCommand;
+     std::shared_ptr<IncrementArmPresetPositionCommand> raiseArmCommand;
+     std::shared_ptr<IncrementArmPresetPositionCommand> lowerArmCommand;
```
Comment (jkleiber) [inline src/main/include/RobotContainer.h:140]: indentation

File: src/test/cpp/subsystems/swerve_drive_test.cpp [hunk 8]
```diff
@@ -1,165 +0,0 @@
-
```
Comment (jkleiber) [inline src/test/cpp/subsystems/swerve_drive_test.cpp:1]: why was this deleted? This shouldn't be deleted

File: src/test/cpp/subsystems/parameters_test.cpp [hunk 9]
```diff
@@ -1,36 +0,0 @@
-
```
Comment (jkleiber) [inline src/test/cpp/subsystems/parameters_test.cpp:1]: Why was this file deleted? It needs to be restored

File: src/test/cpp/utils_test.cpp [hunk 10]
```diff
@@ -1,35 +0,0 @@
-
```
Comment (jkleiber) [inline src/test/cpp/utils_test.cpp:1]: Why is this deleted?

**Review by jkleiber — CHANGES_REQUESTED** [review]
First pass at the PR. Mostly minor comments, but also why are most of the unit tests deleted?

**Comment by danielbrownmsm** [comment]
> First pass at the PR. Mostly minor comments, but also why are most of the unit tests deleted?

um, good question. I think we were working on adding unit tests for the arms, and with the others it was taking longer to build. Or something like that. They might have also been failing? We should add them back, yeah

File: src/main/deploy/parameters.toml (hunk 6 above)
Comment (danielbrownmsm) [reply src/main/deploy/parameters.toml]: after testing it tonight, yeah actually. It's very slow to catch up to the setpoint right now--I think I have an idea on what to do about that that I'll write up in mattermost or something

File: src/main/cpp/io/ArmIO.cpp [hunk 11]
```diff
@@ -32,23 +32,83 @@ bool ArmIO::ProcessIO() {
...
     OKC_CHECK(hw_interface_->arm_extend_motor != nullptr);
     OKC_CHECK(hw_interface_->arm_up_motor != nullptr);
-    hw_interface_->arm_lift_motor->Set(sw_interface_->arm_lift_power);
-    hw_interface_->arm_up_motor->Set(sw_interface_->arm_up_power);
-    hw_interface_->arm_extend_motor->Set(sw_interface_->arm_extend_power);
+   
+    sw_interface_->arm_encoder = hw_interface_->arm_lift_encoder->GetPosition();
+    sw_interface_->arm_extend_encoder = hw_interface_->arm_extend_encoder->GetPosition();
+    sw_interface_->arm_duty_cycle_encoder = hw_interface_->arm_duty_cycle_encoder->GetAbsolutePosition() * 360  -  330; // offset of 330
+
```
[the remark this answers is not in the document]
Comment (danielbrownmsm) [reply src/main/cpp/io/ArmIO.cpp:42]: yes

File: src/main/cpp/io/ArmIO.cpp [hunk 12]
```diff
@@ -32,23 +32,83 @@ bool ArmIO::ProcessIO() {
...
+    if (sw_interface_->arm_duty_cycle_encoder < 0) {
+        sw_interface_->arm_duty_cycle_encoder += 360;
+        sw_interface_->arm_duty_cycle_encoder *= -1;
+    }
+    
+    TeamOKC::Clamp(-0.2, 0.2, &sw_interface_->arm_lift_power);
+    TeamOKC::Clamp(-0.2, 0.2, &sw_interface_->arm_extend_power);
```
[the remark this answers is not in the document]
Comment (danielbrownmsm) [reply src/main/cpp/io/ArmIO.cpp]: yes

File: src/main/cpp/io/ArmIO.cpp (hunk 1 above)
Comment (danielbrownmsm) [reply src/main/cpp/io/ArmIO.cpp]: I don't think we'll end up using an I-term, but yes. I think we should also do some PID setpoint limiting to the actual hardware (and soft) limits, because I was noticing that holding down the arm buttons can get your setpoint really high past the limits really quickly, which isn't good. I might write something in Mattermost up.

File: src/main/cpp/io/ArmIO.cpp (hunk 2 above)
Comment (danielbrownmsm) [reply src/main/cpp/io/ArmIO.cpp]: yeah probably, although I don't necessarily expect it to change

File: src/main/cpp/subsystems/Arm.cpp (hunk 4 above)
Comment (danielbrownmsm) [reply src/main/cpp/subsystems/Arm.cpp]: like AutoArm() or something like we have for Manual? . . . yeah definitely

File: src/main/cpp/commands/arm/IncrementArmExtendCommand.cpp [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+IncrementArmExtendCommand::IncrementArmExtendCommand(std::shared_ptr<Arm> arm,
+                                   double extend) {
+    // Set everything.
+    arm_ = arm;
+    extend_ = extend;
+
+    if (arm_ != nullptr) {
+        this->AddRequirements(arm_.get());
+    
+        arm_->SetControlMode(Auto);
```
[the remark this answers is not in the document]
Comment (danielbrownmsm) [reply src/main/cpp/commands/arm/IncrementArmExtendCommand.cpp:18]: add requirements or SetControlMode? tbh, because the command is running once, we could probably just do it in execute()

File: src/main/cpp/RobotContainer.cpp [hunk 14]
```diff
@@ -197,8 +213,23 @@ bool RobotContainer::InitCommands() {
...
+   
+   
+   
+   //button bindings
+    WPI_IGNORE_DEPRECATED
+    driver_a_button_->WhileHeld(*extendArmCommand);
+    driver_b_button_->WhileHeld(*retractArmCommand);
+    driver_back_button_->WhileHeld(*raiseArmCommand);
+    driver_x_button_->WhileHeld(*lowerArmCommand);
+    WPI_UNIGNORE_DEPRECATED
```
[the remark this answers is not in the document]
Comment (danielbrownmsm) [reply src/main/cpp/RobotContainer.cpp:232]: . . . . yes. I forgot about that. But that's where the ManualArmCommand was so that's where I through it.

File: src/main/cpp/commands/arm/ManualArmCommand.cpp [hunk 15]
```diff
@@ -21,14 +21,11 @@ void ManualArmCommand::Execute() {
 
    
     double lift_power_ = this->gamepad_->GetRawAxis(1);
-    double up_power_ = this->gamepad_->GetRawAxis(0);
-    double extend_power_ = this->gamepad_->GetRawAxis(2);
+    double extend_power_ = this->gamepad_->GetRawAxis(3) - this->gamepad_->GetRawAxis(2);
```
[the remark this answers is not in the document]
Comment (danielbrownmsm) [reply src/main/cpp/commands/arm/ManualArmCommand.cpp:24]: triggers, for manual mode

[the remark this answers is not in the document]
Comment (danielbrownmsm) [reply src/main/cpp/commands/arm/ManualArmCommand.cpp:24]: like, the triggers on the joystick are axes, with a range of [0, 1]

File: src/main/cpp/io/ArmIO.cpp (hunk 3 above)
Comment (danielbrownmsm) [reply src/main/cpp/io/ArmIO.cpp:95]: probably? during testing it was actually the perfect number somehow, but yeah it should be somewhere else

File: src/main/cpp/RobotContainer.cpp [hunk 16]
```diff
@@ -195,10 +220,15 @@ bool RobotContainer::InitCommands() {
...
     swerve_teleop_command_ = std::make_shared<TeleOpSwerveCommand>(swerve_drive_, gamepad1_);
 
-    manual_arm_command_ =std::make_shared<ManualArmCommand>(arm_, gamepad2_);
-    arm_->SetDefaultCommand(*manual_arm_command_);
+    // arm commands
+    extendArmCommand = std::make_shared<IncrementArmExtendCommand>(arm_, 3); 
+    retractArmCommand = std::make_shared<IncrementArmExtendCommand>(arm_, -3);
 
+    raiseArmCommand = std::make_shared<IncrementArmPresetPositionCommand>(arm_, 3);
+    lowerArmCommand = std::make_shared<IncrementArmPresetPositionCommand>(arm_, -3);
```
Comment (jkleiber) [inline src/main/cpp/RobotContainer.cpp:231]: In the interest of making progress, let's add these to the aforementioned ticket about making these arm numbers into parameters

**Review by jkleiber — APPROVED** [review]
You can merge this, but we should parameterize those things I just found in this review eventually if they are things that should be tuned

File: src/main/cpp/RobotContainer.cpp [hunk 17]
```diff
@@ -129,7 +139,13 @@ bool RobotContainer::InitSensors(const Actuators &actuators,
     OKC_CHECK(sensor_interface->right_front_steer_encoder != nullptr);
     OKC_CHECK(sensor_interface->right_back_steer_encoder != nullptr);
 
+    OKC_CHECK(actuators.arm_lift_motor != nullptr);
+    OKC_CHECK(actuators.arm_extend_motor != nullptr);
+
     sensor_interface->arm_lift_encoder = std::make_unique<rev::SparkMaxRelativeEncoder>(actuators.arm_lift_motor->GetEncoder());
+    sensor_interface->arm_duty_cycle_encoder = std::make_unique<frc::DutyCycleEncoder>(1);
+    sensor_interface->arm_extend_encoder = std::make_unique<rev::SparkMaxRelativeEncoder>(actuators.arm_extend_motor->GetEncoder());
+    sensor_interface->extend_limit_switch = std::make_unique<frc::DigitalInput>(0);
```
[the remark this answers is not in the document]
Comment (danielbrownmsm) [reply src/main/cpp/RobotContainer.cpp:148]: I can `#define` those

File: src/main/cpp/RobotContainer.cpp (hunk 16 above)
Comment (danielbrownmsm) [reply src/main/cpp/RobotContainer.cpp:231]: #45
